pkgs/development/libraries/libutempter/default.nix

```diff
@@ -26,6 +27,17 @@ stdenv.mkDerivation rec {
 
   meta = {
     description = "Interface for terminal emulators such as screen and xterm to record user sessions to utmp and wtmp files";
+    longDescription = ''
+    Requires a wrapper to work.
```

Comment:
This explanation only makes sense for NixOS systems. Please try to document the more general requirement (also for non-NixOS systems):  `utempter` must be run as a user belonging to group `utmp`. On NixOS systems, this can be achieved by creating a setuid wrapper...

Reply:
I have thought about this and I think it has more implications than just documentation.

You see, I'm patching the path of utempter from `/var/lib/utempter/utempter` to `/run/wrappers/bin/utempter`. So, as it currently stands, there would be a file with that path required, no matter if on NixOS or any other operating system.

So this might break other systems that had `/var/lib/utempter/utempter` present, maybe as part of the original operating system.
So I guess, I should also make another modification that this path is only patched for NixOs targets. Would this be the preferred way to go?

Reply:
> You see, I'm patching the path of utempter from /var/lib/utempter/utempter to /run/wrappers/bin/utempter. 

Correct, I missed that. `libutempter` should not be patched at all. nixpkgs packages must be self-contained and work on all supported systems (NixOs or other) without any implicit assumptions like the existence of such wrappers.

> So I guess, I should also make another modification that this path is only patched for NixOs targets. Would this be the preferred way to go?

No, as explained above. But maybe you can patch `mosh` to search for `utempter` in `PATH`. That would work on both NixOS and non-nixOS systems: on NixOS systems, `/run/wrappers/bin` is in PATH, and on non-NixOS system the user can add a setgit'd copy of utempter to their PATH. But you need to be able to install and use the `mosh` packages with or without the corresponding module.

Reply:
The path to the `utempter` binary is hard-coded (in libutempter itself, not in mosh) to a path that is usually not part of $PATH on other systems. So, I _could_ patch libutempter to look in $PATH, and that would work on NixOS, but that again would fail on other systems, because on them the binary would be found on a OS-specific path like `/var/lib/utempter/utempter`, which is  usuallynot found on PATH :/

I fail to see a way that this will work in NixOS and other OSes. The way it currently is, it will work most likely in other OSes, but only if they didn't change the path to the utempter binary on their distribution, and libutempter is installed.

Reply:
> The path to the utempter binary is hard-coded (in libutempter itself, not in mosh) to a path that is usually not part of $PATH on other systems. 

Looking at the libutempter source code and derivation, only a _default_ path is hardcoded to `${out}/lib/utempter/utempter` which is in the nix store and will not work on any system.  

But `libutempter` provides an API function `utempter_set_helper()` to set a path to the helper binary. The application must call it to set a path to a setgid version of that helper (or setgid wrapper) because the default path is useless. So you have two options: 

1. patch `mosh` to call `utempter_set_helper()` in a way that works on both NixOS and other systems.

2. patch `libutempter` to look for the helper binary in PATH (or at a sensible location outside the nix store) by default.

The only other nixpkgs package currently using `libutempter` is `guake`. It doesn't do any related patching, so (without having tested it) I assume `utempter` doesn't work with it anyway.  But if you patch `libutempter` please verify that you don't break `guake`.

Anyway, it's up to you to decide if the `utmp` functionality is worth the effort.

Reply:
I guess I'll undo the patch to `libutempter`, modify the code of `mosh` to call `utempter_set_helper`to use `utempter` from the PATH, and wrap the generated `mosh` binary, adding all three `/run/wrappers/bin/utempter`, `/var/lib/utempter/utempter`and `${out}/lib/utempter/utempter` (in that priority) to PATH.
I guess that should cover as many bases as possible.

But I'll be gone for the week, so I'll have to continue this the week after. So expect this to take a while.